community/graphdb-api/src/main/java/org/neo4j/graphdb/PropertyContainer.java

```diff
@@ -145,4 +145,15 @@
      * @return all properties on this property container
      */
     Map<String, Object> getAllProperties();
+
+    /**
+     * Returns the unique id of this <code>PropertyContainer</code>. Ids are garbage collected over time
+     * so they are only guaranteed to be unique during a specific time span: if
+     * the  <code>PropertyContainer</code> is deleted, it's likely that a new
+     * <code>PropertyContainer</code> at some point will get the old id. <b>Note</b>: this makes
+     * <code>PropertyContainer</code> ids brittle as public APIs.
+     *
+     * @return The id of this  <code>PropertyContainer</code>.
+     */
+    long getId();
```

Comment:
You can't actually just define a new method on an interface like this, since it breaks compatibility with implementors. In fact in Neo4j itself there are more implementors of `PropertyContainer` than just `Node` and `Relationship` sub-types.

It needs to be a `default` method on the interface, and I think it should throw `UnsupportedOperationException` in its default implementation. This of course also needs to be documented; that it may throw UOE if the entity in question does not have an Id.